### PR TITLE
lib: add validation package + DNS label validation

### PR DESCRIPTION
**Pull request by eculver**
For future work with mesh federation, we want to validate peer names as valid RFC 1123 DNS labels. This validation already exists, but only in Enterprise and we need the functionality in both OSS and Enterprise. So, I pulled it out and put it into a standalone package under `lib`.

Note: the regex and tests are the same from the original code in Enterprise.

Ask for reviewers: Is this an ok place to put this? I'm also open to renaming things if it makes sense. I'll also link to the PRs in Enterprise that make use of this package and refactor the previous call sites to use this interface. 

Notes to reviewers:
* The calling code will show up in the future PRs that I alluded to above.
* I decided to add both an `IsValidDNSLabel` and a `RequireValidDNSLabel` to a) preserve existing `bool` return + error messages and b) also allow validation rule to be returned to callers if need be -- this is how the mesh federation code does it.

### CODE REVIEW

File: lib/validation/validation.go
Comment (kisunji) [inline lib/validation/validation.go:4]: I think we can keep the comment terse here
```suggestion
// Package validation provides standalone utility functions 
// for validating inputs. 
```

**Review by kisunji — COMMENTED** [review]
Assuming `validation` can grow to mean `anything` maybe let's rename the file `lib/validation/dns.go`?

**Comment by freddygv** [comment]
I left a comment in the mirrored ent PR, I think it could be useful to merge this into the `agent/dns` package, since there's a `InvalidNameRe` there. Not sure why a separate pattern was added for tenancies.

File: .changelog/12535.txt
Comment (freddygv) [inline .changelog/12535.txt:2]: I don't think this should have a user-facing changelog entry, particularly given that the validation from this PR isn't being used yet

**Review by freddygv — COMMENTED** [review]
Thanks for moving this! Have two small requests

File: agent/dns/validation_test.go
Comment (eculver) [reply agent/dns/validation_test.go:22]: Sure, sounds good.

Comment (eculver) [reply agent/dns/validation_test.go:22]: Hah, this is actually really interesting. I pulled the pattern from the existing code in Enterprise and looking at it more closely, it clearly allows 64 characters when it should only allow 63 [according to the RFC](https://datatracker.ietf.org/doc/html/rfc1123#section-2).

So, we just fixed a very nuanced bug. Good catch!

Comment (eculver) [reply agent/dns/validation_test.go:22]: Fixed now, but I want to leave this comment exposed for posterity since it's so fun.

**Review by freddygv — APPROVED** [review]
LGTM assuming CI passes

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/608309.

**Pull request merged**
